src/main/scala/nodes/util/AllSparseFeatures.scala

```diff
@@ -8,12 +8,20 @@ import scala.reflect.ClassTag
 
 /**
  * An Estimator that chooses all sparse features observed when training,
- * and produces a transformer which builds a sparse vector out of them
+ * and produces a transformer which builds a sparse vector out of them.
+ *
+ * Deterministically orders the feature mappings by earliest appearance in the RDD
  */
 case class AllSparseFeatures[T: ClassTag]() extends Estimator[Seq[(T, Double)], SparseVector[Double]] {
   override def fit(data: RDD[Seq[(T, Double)]]): SparseFeatureVectorizer[T] = {
-    val featureSpace = data.flatMap(_.map(_._1)).distinct()
-        .zipWithIndex().collect().map(x => (x._1, x._2.toInt)).toMap
+    val featureOccurrences = data.flatMap(_.map(_._1))
+    // zip with unique ids and take the smallest unique id for a given feature to get
+    // a deterministic ordering
+    val featuresWithUniqueId = featureOccurrences.zipWithUniqueId().reduceByKey {
+      (x, y) => Math.min(x, y)
```

Comment:
What if we make `T <:< Ordering` (or something) and then just do something like
```
 val featureSpace = data.flatMap(_.map(_._1)).distinct()
        .collect().sorted.zipWithIndex.map(x => (x._1, x._2.toInt)).toMap
```
We pay a local sort, but no shuffle, and I don't know how much this is used in practice.